### ISSUE TITLE
Indent JSX in Emacs 27 outside js2-minor-mode

### ISSUE THREAD

**Issue by jacksonrayhamilton** [mooz/js2-mode#529]
JSX indentation is currently only available if

- the user is using Emacs <26 and enables js2-jsx-mode (and this indentation is rather poor), or
- the user is using Emacs >=27 and uses js-mode with js2-minor-mode enabled (this indentation is better), or
- the user installs [rjsx-mode](https://github.com/felipeochoa/rjsx-mode).

It would be nice if in Emacs 27, JSX indentation would be available in plain old js2-mode as well.

Also, currently in Emacs 27, js2-jsx-mode no longer provides any indentation support (since indentation in Emacs 27 relies on js-mode’s syntax-propertize-function, which JS2 does not use; we temporarily decided against that in https://github.com/mooz/js2-mode/pull/523).  This is a regression.

If js2-mode were made to use js-mode’s syntax-propertize-function instead of setting text properties while parsing the AST, then JSX indentation should work again.  This issue proposes making that change.

Before we go forward with that, we’d like to ensure that this change will be compatible with derivative modes (like @felipeochoa’s rjsx-mode) which set syntax-table properties using the AST.  Of particular interest is [this line](https://github.com/felipeochoa/rjsx-mode/blob/03dd8d1683501e81b58674d64c3032b7b718402c/rjsx-mode.el#L781-L782) from rjsx-mode.

I’ll note that an alternative course of action would be to first implement https://github.com/mooz/js2-mode/issues/527 and then set text properties for indentation during the parse of the AST, like we currently do while parsing regexps.  But using syntax-propertize would probably be better for performance and simplicity.

**Comment by felipeochoa** [comment]
If I recall correctly, that text property is set to avoid issues with
unmatched quotes in JSX strings. In general I think it's the "correct"
thing to do there, but possibly duplicative with what js-more is now doing.

-Felipe

**Comment by dgutov** [comment]
I think, if `rjsx-mode` is used (or merged, in any shape or form), the relevant code could be well-positioned to apply the necesarry syntax-table properties by itself.

The ones that make `js-indent-line` work with JSX in Emacs 27.

**Comment by jacksonrayhamilton** [comment]
> I think, if rjsx-mode is used (or merged, in any shape or form), the relevant code could be well-positioned to apply the necesarry syntax-table properties by itself.
> 
> The ones that make js-indent-line work with JSX in Emacs 27.

Yeah, that’s the alternative course of action I referred to.  If we do that, there’s the potential for the properties to be applied more accurately and in better alignment with the AST, at the expense of needing a full parse to figure out where to put the properties.

syntax-propertize could be faster, being chunk-based.  (Then again, it could also slow things down by adding more processing overhead overall on every keystroke.)  Also, it could be easier to just delegate to js-mode rather than write more code.  Decisions, decisions.

**Comment by dgutov** [comment]
> at the expense of needing a full parse to figure out where to put the properties

The full parse will happen anyway, because we want the AST to be available.

> syntax-propertize could be faster, being chunk-based

Could be, in certain respects. However, it's only available in Emacs 27. Even just for backward compatibility reasons, I think we should wait in adopting this approach until 27 is released and widely available.

**Comment by jacksonrayhamilton** [comment]
> Could be, in certain respects. However, it's only available in Emacs 27. Even just for backward compatibility reasons, I think we should wait in adopting this approach until 27 is released and widely available.

Yeah, this is a good argument for using the AST.  I’ll update the OP to reflect this.
